stdlib/public/core/KeyPath.swift

```diff
@@ -3336,7 +3483,14 @@ internal struct InstantiateKeyPathBuffer: KeyPathPatternVisitor {
                                                 mutable: mutable,
                                                 inlineOffset: value)
         pushDest(header)
+        switch kind {
+          case .struct:
+            structOffset += value
+          default:
+             break
```

Comment:
If this is a `.class` component kind, then the struct offset should be invalidated too.

Reply:
I have, around line 3464:
```
    switch kind {
        case .struct:
      isPureStruct.append(true)
        default:
      isPureStruct.append(false)
    }
```
So that looks like it's being taken care of there just so I don't have to do `isPureStruct.append(false)` for every single case. That said, I could remove `isPureStruct.append(false)` on lines 3493 and 3518 since it's being taken care of by the switch statement right at the top of `visitStoredComponent().`

Reply:
That sounds good. As a follow-up, it would be interesting to try to do the optimization for the `outOfLine` and `unresolvedIndirectOffset` cases as well once the final offset is resolved, but we don't need to do that right away.

Reply:
Great! I do have a couple ideas for additional optimizations that'll go in subsequent MRs:
1. The optimization listed in the `Alternative Considered` section at the top.
2. Omit projection in the case where the final N elements are pure structs. For example, if the root is a reference type, and all you have are nested structs down below, we just project from the root to the first struct and then jump to the requisite offset. 

If we include the `unresolvedIndirectOffset` / `outOfLine` optimization, where do you see the best bang-for-the-buck being at this point?

Reply:
Being able to also optimize a tail of struct-offset components sounds promising; you could combine that with a class stored property component either as the final component or as the component before the struct-offset suffix. But if I were going to look into optimizing key path applications myself, I would look at the generic traversal loops and see how we could improve their performance. I suspect that the way they're written with heavy reliance on nested functions and `_openExistential` is not very friendly to the SIL optimizer, and there's a lot of unneeded overhead that could be avoided by an optimal implementation that eliminated unnecessary copies and retain/release.

Reply:
Sounds good! I have the benchmarks for the struct-offset-tail complete, and the actual implementation is in the debugging phase. Getting these up is something I'll be working on next.

What other methods of getting concrete type information inside a scope, other than `_openExistential` exist, if any? I did at one point float the idea of grabbing the concrete type information via `_openExistential` and storing it for subsequent read and write operations, but now it looks like that would have a bigger impact on ABI stability than we might like.

Reply:
There should not be any ABI concerns with changing the internal implementation of `swift_getKeyPath` or its variants, since their implementations are private. A blunt approach would be to rewrite those functions in runtime C++, where we have finer control over access to type metadata and value copying. `_openExistential` shouldn't really incur any overhead directly, but I suspect that the functions we open into aren't always getting inlined, and that may be introducing overhead. I would look at the quality of the generated code and see what combination of compiler improvements and/or code changes we can make to reduce overhead there.

Reply:
Sounds like a plan, thanks!